src/mlpack/core/tree/cosine_tree/cosine_tree.cpp

```diff
@@ -150,6 +150,196 @@ CosineTree::CosineTree(const arma::mat& dataset,
   ConstructBasis(treeQueue);
 }
 
+//! Copy the given tree.
+CosineTree::CosineTree(const CosineTree& other) :
+    dataset(other.dataset),
```

Comment:
The same copying issue as in @walragatver's comment will apply here too. :+1:

Reply:
I'm not entirely sure if these are the fixes you meant for but they seem correct 🙂

Reply:
Hmm they definitely aren't correct. Could you go into a bit more detail about how I should be handling the deletion of the `dataset` pointer?

Reply:
Well, so there are a couple options.  The one I suggested is to hold a `bool` inside the class indicating whether or not the `CosineTree` owns the `dataset` pointer (and thus is responsible for calling `delete` on it in the destructor).

The other approach you could take is to adapt the `CosineTree` to be more like, e.g., the `BinarySpaceTree`, where the root node *always* owns the dataset and must free it in the destructor.  That will be slightly more work though since you will have to adapt the code that uses the `CosineTree`.

Reply:
> Well, so there are a couple options. The one I suggested is to hold a `bool` inside the class indicating whether or not the `CosineTree` owns the `dataset` pointer (and thus is responsible for calling `delete` on it in the destructor).

Ah I see. Like the `localDataset` variable in one of the other trees. Will do.

Reply:
This still has a problem.  Suppose that I do the following code:

```
arma::mat dataset(100, 100, arma::fill::randu);
CosineTree* ct = new CosineTree(dataset);
CosineTree copy(*ct); // make a copy of `ct`
delete ct; // deletes `ct.dataset`
// now copy.dataset is invalid!
```

So the copy constructor should copy everything---including the dataset.  Therefore this particular line will need to be updated.

Reply:
https://pastebin.com/rP7kZeQc

I ran the above sample and it worked fine. What do you think? I understand what you said but it works regardless

Reply:
Deep copies aren't being made when the pointer is copied. Try running that code with `valgrind` and you should see that there is an invalid memory access. :+1:

Reply:
Could you take a look now @rcurtin ?

Reply:
Ah, I didn't see the edit when I responded this morning:

> Edit2: So I ran a few more tests and I found out a few things. A deep copy isn't taking place (which should have been obvious actually 😅). copy.dataset tree points to the same location as ct->dataset. Calling delete ct sets ct->dataset to NULL but does not delete the dataset variable itself, allowing the copied tree to access the dataset.
>
> I assume this is because the dataset variable is allocated on the stack and thus not deleted. Is this okay or should I be deleting it somehow?

Ah I realized there is a problem in my example code.  Here we actually need to do this:

```
arma::mat* dataset = new arma::mat(100, 100, arma::fill::randu);
CosineTree* ct = new CosineTree(dataset);
CosineTree copy(*ct); // make a copy of `ct`
delete dataset;
// now copy.dataset is invalid!
```

Basically the issue with my previous code snippet was that `ct` does not "own" `ct->dataset` so it will not delete it.  If `ct` *does* own `ct->dataset`, or if `dataset` is owned by something else and then deleted, then `copy.dataset` will be invalid.  So when we use the copy constructor, we need to make sure that the `CosineTree` we're creating will not have this problem, and that includes copying the dataset.  Let me know if I can clarify that.

Reply:
Oh okay I get it, that makes sense